src/node/miner.cpp

```diff
@@ -446,4 +449,94 @@ void AddMerkleRootAndCoinbase(CBlock& block, CTransactionRef coinbase, uint32_t
     block.nNonce = nonce;
     block.hashMerkleRoot = BlockMerkleRoot(block);
 }
+
+std::optional<std::unique_ptr<CBlockTemplate>> WaitNext(NodeContext& node, const uint256& prev_block_hash, const std::vector<CAmount>& tx_fees, const BlockWaitOptions& options, const BlockAssembler::Options& assemble_options)
```

Comment:
In commit "interfaces: refactor: move `waitNext` implementation to miner" (a5145ba0fc3339392ff71bf20613812e0d8c8efc)

This change generally looks good, but I would suggest tweaking this commit and the next one to pass Chainman& and KernelNotification& parameters directly instead of NodeContext& because NodeContext exposes a lot of node state that this code shouldn't need access to and also brings in dependencies this code shouldn't need to depend on.

Reply:
Thanks for taking a look
I've taken your suggestion with the addition of `CTxMemPool*` as a parameter to `WaitNext`.
I think it's best to pass it as well, to make the method self contained and allow for testing with a custom mempool.

Also, if I use the mempool pointer in the active chainstate not the node, a miner test fails.
This happens because in `TestPackageSelection`, we construct a new mempool and update the node interfaces' mempool pointer, but we don't update the raw pointer stored in the active chainstate.